### PR TITLE
Rust: Don't consider parameters in trait method definitions without bodies as variables

### PR DESCRIPTION
**Pull request by paldepind**
This PR fixes false positives in the `rust/unused-value` and `rust/unused-variable` queries for parameters in trait methods.

I decided to make the change in the variable implementation, as that fixes both queries in one swoop and I don't think there would ever be a need to have these parameters considered as variables by the variable implementation.

It's more illuminating to look at the commits individually, as the composite of the diffs kinda obscures how the results changes.

### CODE REVIEW

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll [hunk 1]
```diff
@@ -84,7 +84,10 @@ module Impl {
     // an enum constant (e.g. `None`). This excludes static and constant variables (UPPERCASE),
     // which we don't appear to recognize yet anyway. This also assumes programmers follow the
     // naming guidelines, which they generally do, but they're not enforced.
-    not name.charAt(0).isUppercase()
+    not name.charAt(0).isUppercase() and
+    // exclude parameters from functions without a bodies as these are trait method declarations
+    // without implementations
+    not exists(Function f | not f.hasBody() and f.getParamList().getAParam().getPat() = p)
```
Comment (paldepind) [inline rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:90]: Note that this just compares the `IdentPat` directly with the pattern from the parameter, whereas the `getParameter` predicate does some more complicated work to get a parameter. However, patterns aren't allowed in parameters for functions without bodies, so I think this should be sufficient in this case.

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll [hunk 2]
```diff
@@ -84,7 +84,10 @@ module Impl {
     // an enum constant (e.g. `None`). This excludes static and constant variables (UPPERCASE),
     // which we don't appear to recognize yet anyway. This also assumes programmers follow the
     // naming guidelines, which they generally do, but they're not enforced.
-    not name.charAt(0).isUppercase()
+    not name.charAt(0).isUppercase() and
+    // exclude parameters from functions without a bodies as these are trait method declarations
```
Comment (redsun82) [inline rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:88]:
```suggestion
    // exclude parameters from functions without a body as these are trait method declarations
```

**Review by redsun82 — DISMISSED** [review]
LGTM, just one small typo in a comment 🙂

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 2 above)
Comment (paldepind) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll]: Thanks! I've applied the change.

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 1 above)
Comment (aibaars) [inline rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:90]: I'm not sure this is entirely correct. What about tuple parameters. For example:
```rust
fn foo(point@(x, y): (i32, i32)) { ... }
```

I'm also not entirely sure we don't want those parameters to be variables. They might be useful for dataflow when binding arguments to parameters in a function call.

Comment (redsun82) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:90]: I'd say we bring this in as is and leave this for a follow up though. I think we can already a good chunk of FPs with this.

**Review by redsun82 — APPROVED** [review]
Let's leave the valid remark by @aibaars for a follow up

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 1 above)
Comment (redsun82) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:90]: ah, wait, thinking more clearly about it, yeah, maybe it's better not to put this restriction at the `VariableImpl` level, but something more specific for the unused variable/value query implementations.

Comment (paldepind) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:90]: @aibaars See the comment I wrote here: https://github.com/github/codeql/pull/17847#discussion_r1816762685. This is only done for functions without bodies and patterns are not allowed for parameters in such functions. Rust will error for your example if you remove the body and put it in a trait.

> I'm also not entirely sure we don't want those parameters to be variables. They might be useful for dataflow when binding arguments to parameters in a function call.

Also for functions that have no body? If that is the case we should do the similar thing but directly in the two queries.

**Comment by paldepind** [comment]
> Let's leave the valid remark by @aibaars for a follow up

Alright 👍  If we decide on another approach I'll make a follow up PR.

File: rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll (hunk 1 above)
Comment (aibaars) [reply rust/ql/lib/codeql/rust/elements/internal/VariableImpl.qll:90]: Ah, didn't spot that comment, in that case it should be fine.

Whether or not it is useful to consider the parameters of trait functions depends a bit on how we connect the arguments to parameters. If we first connect the arguments of a call to the parameters of the function defined in a trait and subsequently link them to all suitable `impl`s of the function then we might end up wondering why nothing works ;-)  I guess if you directly link the arguments of a call to all suitable `impl`s then things would just work. Just keep this in mind when implementing dataflow calls.

**Pull request merged**